meta-xt-domu/recipes-gnome/gtk+/gtk+3_%.bbappend

```diff
@@ -0,0 +1,4 @@
+# Avoid installation of the redundant 'adwaita-icon-theme' package.
+# It is causing build of the rust, which we want to avoid, as it takes a lot of time.
+RRECOMMENDS:${PN}:remove = " adwaita-icon-theme-symbolic"
```

Comment:
I am not sure that this is the correct thing to do. Okay, maybe **we** don't need this packages. But someone of our users might need them. So, this change should not be present in `meta-xt-common`.

Maybe we can add something similar into product layer... But it definitely should not be there.

Reply:
Hi @lorc, would it be sufficient, if I'll move these changes to meta-layers of [meta-xt-prod-devel-rcar](https://github.com/xen-troops/meta-xt-prod-devel-rcar)?

If yes, which of the existing meta-layers there would fit the purpose?

|Domain name|
|meta-agl-tune|
|meta-aos-tune|
|meta-xt-domx-gen3|
|meta-xt-prod-devel-rcar-control|
|meta-xt-prod-devel-rcar-driver-domain|

I can move what was in the 'meta-xt-driver-domain' layer in this commit into the 'meta-xt-prod-devel-rcar-driver-domain'.

But what about the content of the meta-xt-domu? Can I create a new meta-layer with such a name in the meta-xt-prod-devel-rcar?

Reply:
> Hi @lorc, would it be sufficient, if I'll move these changes to meta-layers of [meta-xt-prod-devel-rcar](https://github.com/xen-troops/meta-xt-prod-devel-rcar)?

Yes, it will be better.

> But what about the content of the meta-xt-domu? 

There is `meta-xt-domx-gen3` for recipes that are used in more than one domain

Reply:
Hi @lorc, then agreed. I'll move it to [meta-xt-prod-devel-rcar](https://github.com/xen-troops/meta-xt-prod-devel-rcar) next week. Thanks for your feedback.

Reply:
Hi @lorc, I've started to apply this change. But I've faced the following issue.

I need to apply this change only to DomD and DomU. No other domains should be impacted.

But, according to the moulin yaml file, the "meta-xt-domx-gen3" layer is used for DomD and Dom0. And not for DomU.

As a result, your proposal to place it inside the "meta-xt-domx-gen3" will impact DomD and Dom0, without impacting DomU. That is not what was intended by this change.

Moreover, I've found that DomU does not use any layers from meta-xt-prod-devel-rcar:

```
layers:
  - "../meta-openembedded/meta-oe"
  - "../meta-openembedded/meta-filesystems"
  - "../meta-openembedded/meta-python"
  - "../meta-openembedded/meta-networking"
  - "../meta-openembedded/meta-gnome"
  - "../meta-renesas/meta-rcar-gen3"
  - "../meta-xt-common/meta-xt-domu"
  - "../meta-xt-rcar/meta-xt-rcar-fixups"
  - "../meta-xt-rcar/meta-xt-rcar-domu"
  - "../meta-xt-rcar/meta-xt-rcar-gles_common"
```

I do not see a possibility to move this change to meta-xt-prod-devel-rcar without creating a new domu-specific layer in that repository.

**=> How should we proceed? Leave the change here? Create a new layer in meta-xt-prod-devel-rcar? Or, maybe, you have some other ideas?**

Reply:
Well, this is not carved in the stone. DomU didn't used meta-xt-domx-gen3 because there was no need for this. Now there is a such need, so you can add it to the list of layers. 

In the append you can use `XT_DOM_NAME` variable to determine which domain is being built right now and filter out dom0.

Reply:
Hi @lorc, your proposal might not be sufficient. In my case, we are talking about the e.g. gtk+3_%.bbappend.

If I will place that bbappend file into the meta-xt-domx-gen3 layer, that would mean, that all moulin components,  which are using that layer, will need to use gtk bb recipe. In my case, DomD and DomU have such bitbake recipe for sure. But Dom0 might not have it.

Such a situation will lead to a build failure, as it is not possible to use bbappend without underlying bb used somewhere in the Yocto build.

And XT_DOM_NAME will not allow you to step over it, as it is not possible to use XT_DOM_NAME to filter out which bbappend to use and which not. The proposed approach can be used only for defining conditional logic inside the bbappend file, but not for skipping bbappend file as a whole.

Moreover, meta-xt-domx-gen3 layer already contains xen_git.bbappend, xen-tools_git.bbappend, which are used for Dom0 & DomD. AFAIK DomU does not have xen-tools inside, so it will be totally the same situation.

**=> Any ideas regarding how to step over this issue?**

Reply:
Ah, I see. Right.

Okay, taking into account that we are fighting `RRECOMENDS`, Yocto Manual suggests to use `BAD_RECOMMENDATIONS` variable. Now, where to place it... 

@otyshchenko1 had very similar issue with `qemu` in `xen-tools`, and in that case we agreed to update `RRECOMENDS` in the `xen-tools.bbappend`, which is not perfect, but okayish.

Generally, we want to use `domx` layer in this case, but as you pointed out, it provides some difficulties with the current design. 

As a temporal solution I propose to add `BAD_RECOMENDATIONS:append=" adwaita-icon-theme-symbolic"` into YAML file along with an appropriate comment, of course. Also, please create a github issue describing this problem. We want to track such temporary solutions.

Reply:
Hi @lorc

After a phone call with you, it was decided to:

- Add a new meta-layer to the meta-xt-prod-devel-rcar repository - meta-xt-domd-gen3
- Add a new meta-layer to the meta-xt-prod-devel-rcar repository - meta-xt-domu-gen3
- Assign those 2 layers to the corresponding domains inside the moulin yaml configuration
- Relocate this change to those layers
- Try to use BAD_RECOMENDATIONS instead of the RRECOMMENDS:${PN}:remove
- Try to find out whether it is possible to replace the PACKAGECONFIG:remove directive with turning off some DISTRO_FEATURES

I start to proceed with this implementation.

Reply:
> In both cases 'adwaita-icon-theme-symbolic' and 'rust' are presented in the graph. Dot graphs attached:

I am not surprised at all :) Bitbake has no `:add` syntax. You need to use `:append`.

> rsvg is declared as a PACKAGECONFIG, which is not part of any DISTRO_FEATURES.

> No occurrences, where the availability of rsvg would be caused depending on some DISTRO_FEATURE.

Yes, because it depends on the `TUNE_FEATURES` variable: https://github.com/yoctoproject/poky/blob/master/meta/recipes-multimedia/gstreamer/gstreamer1.0-plugins-bad_1.22.2.bb#L32

Don't know how you missed this. Anyways, investigation continues. You need to understand purpose of `TUNE_FEATURES` variable and where it is set up.

Reply:
Hi @lorc,

Regarding add vs. append - you are right. What a silly typo. I'll try out the correct syntax and will let you know whether it works.

----

Regarding TUNE_FEATURES. I've noticed it right from the beginning. But I doubt that it is something that we can use for the configuration of rsvg. According to the documentation, it means the following:

_[TUNE_FEATURES](https://docs.yoctoproject.org/ref-manual/variables.html#term-TUNE_FEATURES)[](https://docs.yoctoproject.org/ref-manual/variables.html#term-TUNE_FEATURES)
Features used to “tune” a compiler for optimal use given a specific processor. The features are defined within the tune files and allow arguments (i.e. TUNE_*ARGS) to be dynamically generated based on the features.

The OpenEmbedded build system verifies the features to be sure they are not conflicting and that they are supported.

The BitBake configuration file (meta/conf/bitbake.conf) defines [TUNE_FEATURES](https://docs.yoctoproject.org/ref-manual/variables.html#term-TUNE_FEATURES) as follows:

TUNE_FEATURES ??= "${TUNE_FEATURES:tune-${DEFAULTTUNE}}"
See the [DEFAULTTUNE](https://docs.yoctoproject.org/ref-manual/variables.html#term-DEFAULTTUNE) variable for more information._

From my understanding the line "${@bb.utils.contains('TUNE_FEATURES', 'mx32', '', 'rsvg', d)} \" means, that in the case of mx32, the rsvg is not used. In this context "mx32" stands for the GCC compiler option:

```
https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html
The -mx32 option sets int, long, and pointer types to 32 bits, and generates code for the x86-64 architecture.
```

It seems, that it is x86 specific option, which is not available for arm:

https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html

So, for arm rsvg is ALWAYS used. And there is no possibility to use TUNE_FEATURES to adjust this.

Reply:
Hi @lorc 

[BAD_RECOMMENDATIONS](https://docs.yoctoproject.org/3.2.3/ref-manual/ref-variables.html#term-BAD_RECOMMENDATIONS)[](https://docs.yoctoproject.org/3.2.3/ref-manual/ref-variables.html#term-BAD_RECOMMENDATIONS)
Lists “recommended-only” packages to not install. Recommended-only packages are packages installed only through the [RRECOMMENDS](https://docs.yoctoproject.org/3.2.3/ref-manual/ref-variables.html#term-RRECOMMENDS) variable. You can prevent any of these “recommended” packages from being installed by listing them with the BAD_RECOMMENDATIONS variable

**=> BAD_RECOMMENDATIONS option does not impact the build of the target package. It impacts only the installation. But our goal in this change is to minimize build time. So BAD_RECOMMENDATIONS does not fit.**

It seems that I've addressed all the findings. Here is what the target PR would look like:
https://github.com/dterletskiy/meta-xt-prod-devel-rcar/pull/3
It is based on another branch. But diff is totally the same. Let me know whether it is OK from your point of view. If it is - I'll ask @otyshchenko1 to integrate it into the corresponding PR and drop this commit.

Reply:
Okay, fair enough. Thanks for the investigation. So yes, I agree, only one viable option is to use `:remove` syntax for both `RRECOMENDS` and `PACKAGE_CONFIG` in this case

Reply:
"+"